app/code/Magento/ImportExport/Model/Import/Source/FileParser/CompositeParserFactory.php

```diff
@@ -0,0 +1,62 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento\ImportExport\Model\Import\Source\FileParser;
+
+/**
+ * File Parser Composite
+ *
+ */
+class CompositeParserFactory implements ParserFactoryInterface
+{
+    /** @var ParserFactoryInterface[] */
+    private $parserFactories;
+
+    public function __construct(array $parserFactories = [])
+    {
+        $this->parserFactories = [];
+
+        foreach ($parserFactories as $parserFactory) {
+            $this->addParserFactory($parserFactory);
+        }
+    }
+
+    public function create($path, array $options = [])
+    {
+        foreach ($this->parserFactories as $parserFactory) {
+            $parser = $this->createParserIfSupported($parserFactory, $path, $options);
+
+            if ($parser === false) {
+                continue;
+            }
+
+            return $parser;
+        }
+
+        $this->thereWasNoParserFound($path);
+    }
+
+    public function addParserFactory(ParserFactoryInterface $parserFactory)
+    {
+        $this->parserFactories[] = $parserFactory;
+    }
+
+    private function thereWasNoParserFound($path)
+    {
+        throw new UnsupportedPathException($path);
+    }
+
+    private function createParserIfSupported(ParserFactoryInterface $parserFactory, $path, array $options)
+    {
+        try {
+            $parser = $parserFactory->create($path, $options);
+        } catch (UnsupportedPathException $e) {
+            return false;
```

Comment:
Idea is to have a decision point. The decision point in your case is your composite factory. 
The decision point will detect which Factory should be used in each particular case and then just instantiate **the only** needed factory to process the document.
In this case, we will instantiate just one factory, instead all of the factories in advance (even if we need the only one). And will not try to process document with factories which does not fit for this document.

In this case, Exception would be a real exceptional state, not like it's now when **Exception is Expected** behavior.

>  returning false is a terrible thing.

Yes, that's why I am advising you to make your code more Object Oriented friendly to get rid of such code
```php
$parser = $this->createParserIfSupported($parserFactory, $path, $options);
if ($parser === false) {
    continue;
 }
```

_Problems you have with Exceptions and False handling is just a **side-effect of wrong OO design** with choosing right Factory to handle the precise business case_